library/core/src/mem/transmutability.rs

```diff
@@ -13,6 +66,53 @@ pub unsafe trait BikeshedIntrinsicFrom<Src, const ASSUME: Assume = { Assume::NOT
 where
     Src: ?Sized,
 {
+    /// Transmutes a `Src` value into a `Self`.
+    ///
+    /// # Safety
+    ///
+    /// The safety obligations of the caller depend on the value of `ASSUME`:
+    /// - If [`ASSUME.alignment`](Assume::alignment), the caller must prove that
+    ///   all references in `src` (including `src` itself, if `Src` is a
+    ///   reference type) have well-aligned referents upon transmutation into
+    ///   `Self`.
+    /// - If [`ASSUME.lifetimes`](Assume::lifetimes), the caller must prove that
+    ///   all references in `src` (including `src` itself, if `Src` is a
+    ///   reference type) have referents that live as long as `Self`.
+    /// - If [`ASSUME.safety`](Assume::safety), the caller must prove that the
+    ///   value `src` transmuted into `Self` satisfies all library safety
+    ///   invariants of `Self`.
+    /// - If [`ASSUME.validity`](Assume::validity), the caller must prove that
+    ///   `src` is a bit-valid instance of `Self`.
+    ///
+    /// When satisfying the above obligations (if any), the caller must *not*
+    /// assume that this trait provides any inherent guarantee of layout
+    /// [portability](#portability) or [stability](#stability).
+    unsafe fn transmute(src: Src) -> Self
```

Comment:
@compiler-errors, I'm getting an ICE when I try to call this method in tests:
```
---- core/src/mem/transmutability.rs - mem::transmutability::BikeshedIntrinsicFrom (line 41) stdout ----
error: internal compiler error: compiler/rustc_middle/src/ty/instance.rs:599:21: failed to resolve instance for <Dst as BikeshedIntrinsicFrom<u8, Assume { alignment: false, lifetimes: false, safety: true, validity: false }>>::transmute
  --> core/src/mem/transmutability.rs:51:5
   |
13 |     <Dst as BikeshedIntrinsicFrom<u8, { Assume::SAFETY }>>::transmute(src)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

thread 'rustc' panicked at compiler/rustc_middle/src/ty/instance.rs:599:21:
Box<dyn Any>
stack backtrace:
   0:     0x7f11424685b6 - trace
                               at /home/ubuntu/projects/rust/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x7f11424685b6 - trace_unsynchronized<std::sys::backtrace::_print_fmt::{closure_env#1}>
                               at /home/ubuntu/projects/rust/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f11424685b6 - _print_fmt
                               at /home/ubuntu/projects/rust/library/std/src/sys/backtrace.rs:66:9
   3:     0x7f11424685b6 - fmt
                               at /home/ubuntu/projects/rust/library/std/src/sys/backtrace.rs:39:26
   4:     0x7f11424c82eb - fmt
                               at /home/ubuntu/projects/rust/library/core/src/fmt/rt.rs:173:76
[...]
```
https://github.com/rust-lang/rust/blob/80eb5a8e910e5185d47cdefe3732d839c78a5e7e/compiler/rustc_middle/src/ty/instance.rs#L599-L603

Do you know if there any special ceremony needed to add a defaulted method to a compiler-implemented trait?

Reply:
Add a case to the `traits::ImplSource::Builtin(BuiltinImplSource::Misc, _)` arm in `resolve_associated_item` in `compiler/rustc_ty_utils/src/instance.rs`. Look for the `clone_from` case, for example.

Reply:
Thanks! [Done!](https://github.com/rust-lang/rust/pull/129032/files#diff-a54c982c1e255569ddede4404cdf85ecfbd93c20edfa8a090a5f2dc2bc1dff6b)